server/Main.hs

```diff
@@ -53,31 +53,36 @@ instance A.ToJSON Error
 
 server :: [P.ExternsFile] -> P.Env -> P.Environment -> Int -> IO ()
 server externs initNamesEnv initEnv port = do
+  let toCompilerErrors :: NE.NonEmpty CST.ParserError -> Either Error a
+      toCompilerErrors = Left . CompilerErrors . P.toJSONErrors False P.Error . CST.toMultipleErrors "<file>"
+
   let compile :: Text -> IO (Either Error ([P.JSONError], JS))
       compile input
         | T.length input > 20000 = return (Left (OtherError "Please limit your input to 20000 characters"))
         | otherwise = do
-          case CST.parseModuleFromFile "<file>" input >>= CST.resFull of
+          case CST.parseModuleFromFile "<file>" input of
             Left parseError ->
-              return . Left . CompilerErrors . P.toJSONErrors False P.Error $ CST.toMultipleErrors "<file>" parseError
-            Right m | P.getModuleName m == P.ModuleName "Main" -> do
-              (resultMay, ws) <- runLogger' . runExceptT . flip runReaderT P.defaultOptions $ do
-                ((P.Module ss coms moduleName elaborated exps, env), nextVar) <- P.runSupplyT 0 $ do
-                  desugared <- P.desugar initNamesEnv externs [P.importPrim m] >>= \case
-                    [d] -> pure d
-                    _ -> error "desugaring did not produce one module"
-                  P.runCheck' (P.emptyCheckState initEnv) $ P.typeCheckModule desugared
-                regrouped <- P.createBindingGroups moduleName . P.collapseBindingGroups $ elaborated
-                let mod' = P.Module ss coms moduleName regrouped exps
-                    corefn = CF.moduleToCoreFn env mod'
-                    [renamed] = P.renameInModules [corefn]
-                unless (null . CF.moduleForeign $ renamed) . throwError . P.errorMessage $ P.MissingFFIModule moduleName
-                P.evalSupplyT nextVar $ P.prettyPrintJS <$> J.moduleToJs renamed Nothing
-              case resultMay of
-                Left errs -> (return . Left . CompilerErrors . P.toJSONErrors False P.Error) errs
-                Right js -> (return . Right) (P.toJSONErrors False P.Error ws, js)
-            Right _ ->
-              (return . Left . OtherError) "The name of the main module should be Main."
+              return $ toCompilerErrors parseError
+            Right partialResult -> case CST.resFull partialResult of
+              (_, Left parseError) ->
+                return $ toCompilerErrors parseError
+              (warnings, Right m) | P.getModuleName m == P.ModuleName "Main" -> do
+                (resultMay, ws) <- runLogger' . runExceptT . flip runReaderT P.defaultOptions $ do
+                  ((P.Module ss coms moduleName elaborated exps, env), nextVar) <- P.runSupplyT 0 $ do
```

Comment:
This part of the code is basically an inlined version of `Language.PureScript.Make.rebuildModule`, but `rebuildModule` [has changed inside the compiler](https://github.com/purescript/purescript/compare/v0.13.8...v0.14.0-rc5#diff-fae31b1b9fa5fe55802867766a0c2f98487abbaab050fe24fc40edb162e33d9e) since this was first written so I think this is now not quite right. I think previously this was mostly-inlined from the compiler because using rebuildModule directly would require coming up with a `MakeActions` record, and in particular we'd need to provide a `codegen` field which captures the JS output in a mutable variable rather than writing it to disk. I think previously that probably didn't seem worth the effort, but now that `rebuildModule` has become more complex, it's probably better to take this route.

The `Make` monad in Language.PureScript.Make should be an appropriate choice for the monad type variable, and you can use `runMake` like we already did elsewhere in this file to get back down to IO. Let me know if you'd like to pair on this.

Reply:
I'm going to take a stab at this right now, but if I'm stumped then I'd appreciate pairing.

Reply:
There's quite a lot going on in `MakeActions`! I think it would help me to pair on this, but I've stubbed out an implementation using `IORef` for the time being so there's a place to start.